SCAutolib/controller.py

```diff
@@ -0,0 +1,375 @@
+
+        # Add IPA packages if needed
+        if not all([u["local"] for u in self.lib_conf["users"]]):
+            packages += self._general_steps_for_ipa()
+
+        # Check for installed packages
+        missing = _check_packages(packages)
+        if install_missing:
+            _install_packages(missing)
+
+        run(['dnf', 'groupinstall', "Smart Card Support", '-y'])
+        logger.debug("Smart Card Support group in installed.")
+
+        self.sssd_conf.create()
+        self.sssd_conf.save()
+        self._general_steps_for_virtual_sc()
+
+    def setup_local_ca(self, force: bool = False):
+        """
+        Setup local CA based on configuration from the configuration file. All
+        necessary file for this operation (e.g. CNF file for self-signed root
+        certificate) would be created along the way.
+
+        :param force: If local CA already exists in given directory, specifies
+            if it should be overwritten
+        :type force: bool
+        :raises: SCAutolib.exceptions.SCAutolibWrongConfig
+        """
+
+        if "local_ca" not in self.lib_conf["ca"].keys():
+            msg = "Section for local CA is not found in the configuration file"
+            raise SCAutolibWrongConfig(msg)
+
+        ca_dir: Path = self.lib_conf["ca"]["local_ca"]["dir"]
+        cnf = file.OpensslCnf(ca_dir.joinpath("ca.cnf"), "CA", str(ca_dir))
+        self.local_ca = CA.LocalCA(dir=ca_dir, cnf=cnf)
+
+        if force or not ca_dir.exists():
+            logger.warning(f"Removing previous local CA in a directory "
+                           f"{ca_dir}")
+            self.local_ca.cleanup()
+
+            ca_dir.mkdir(exist_ok=True)
+
+            cnf.create()
+            cnf.save()
+            self.local_ca.setup()
+        logger.info(f"Local CA is configured in {ca_dir}")
+        # Generate certificates
+
+    def setup_ipa_client(self, force: bool = False):
+        """
+        Configure IPA client for given IPA server on current host. IPA server
+        should be already up and running for correct configuration of the IPA
+        client
+
+        :param force: If IPA Client is already configured on the system,
+            specifies if it should be removed before configuring a new client.
+        :type force: bool
+        :raises: SCAutolib.exceptions.SCAutolibWrongConfig
+        """
+        if "ipa" not in self.lib_conf["ca"].keys():
+            msg = "Section for IPA is not found in the configuration file"
+            raise SCAutolibWrongConfig(msg)
+        self.ipa_ca = CA.IPAServerCA(**self.lib_conf["ca"]["ipa"])
+
+        if self.ipa_ca.is_installed:
+            logger.warning("IPA client is already configured on this system.")
+            if not force:
+                logger.info("Set force argument to True if you want to remove "
+                            "previous installation.")
+                return
+            self.ipa_ca.restore()
```

Comment:
Why do we need ipa_ca.restore() here?

Reply:
This will remove IPA client installation if `force` option is set